src/Database/LSMTree/Internal.hs

```diff
@@ -439,20 +453,24 @@ closeSession Session{sessionState, sessionTracer} = do
     RW.withWriteAccess_ sessionState $ \case
       SessionClosed -> pure SessionClosed
       SessionOpen seshEnv -> do
+        -- Close tables and cursors first, so that we know none are open when we
+        -- release session-wide resources.
+        --
+        -- If any has been closed already by a different thread, the idempotent
+        -- 'close' will act like a no-op, and so we are not in trouble.
+        --
         -- Since we have a write lock on the session state, we know that no
         -- tables will be added while we are closing the session, and that we
         -- are the only thread currently closing the session.
         --
         -- We technically don't have to overwrite this with an empty Map, but
         -- why not
-        tables <- modifyMVar (sessionOpenTables seshEnv) (\m -> pure (Map.empty, m))
-        -- Close tables first, so that we know none are open when we release
-        -- session-wide resources.
-        --
-        -- If any table from this set has been closed already by a different
-        -- thread, the idempotent 'close' will act like a no-op, and so we are
-        -- not in trouble.
-        mapM_ close tables
+        modifyMVar_ (sessionOpenTables seshEnv) $ \tables -> do
+          mapM_ close tables
+          pure Map.empty
```

Comment:
I think this is slightly nicer, since if there is an error while closing tables, `sessionOpenTables` will keep the old values instead of being empty. Better to close twice than not at all.

Reply:
Ah, but it causes a deadlock, since `close` wants to remove the handle from `sessionOpenTables`: https://github.com/IntersectMBO/lsm-tree/actions/runs/10489456896/job/29054138188?pr=345#step:16:1583

Reply:
Putting back closed tables is not bad, as `close` is idempotent. But I agree that this generally needs `TempRegistry`.

Reply:
It can lead to a memory leak for long-lived programs, because only `sessionClose` would get rid of the closed table from `sessionOpenTables`. It's a contrived scenario, but probably one we should avoid